cpp/include/raft/common/nvtx.hpp

```diff
@@ -0,0 +1,119 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "detail/nvtx.cuh"
+
+namespace raft {
+namespace common {
+
+/**
+ * @brief Push a named nvtx range
+ * @param format range name format (accepts printf-style arguments)
+ * @param args the arguments for the printf-style formatting
+ */
+template <typename... Args>
+inline void PUSH_RANGE(const char* format, Args... args)
+{
+  detail::pushRange(format, args...);
+}
+
+/**
+ * @brief Synchronize CUDA stream and push a named nvtx range
+ * @param format range name format (accepts printf-style arguments)
+ * @param args the arguments for the printf-style formatting
+ * @param stream stream to synchronize
+ */
+template <typename... Args>
+inline void PUSH_RANGE(rmm::cuda_stream_view stream, const char* format, Args... args)
+{
+  detail::pushRange(stream, format, args...);
+}
+
+/** Pop the latest range */
+inline void POP_RANGE() { detail::popRange(); }
+
+/**
+ * @brief Synchronize CUDA stream and pop the latest nvtx range
+ * @param stream stream to synchronize
+ */
+inline void POP_RANGE(rmm::cuda_stream_view stream) { detail::popRange(stream); }
+
+/** Push a named nvtx range that would be popped at the end of the object lifetime. */
+class AUTO_RANGE {
+ private:
+  std::optional<rmm::cuda_stream_view> streamMaybe;
+
+  /* This object is not meant to be touched. */
+  AUTO_RANGE(const AUTO_RANGE&) = delete;
+  AUTO_RANGE(AUTO_RANGE&&)      = delete;
+  AUTO_RANGE& operator=(const AUTO_RANGE&) = delete;
+  AUTO_RANGE& operator=(AUTO_RANGE&&) = delete;
+
+ public:
+  /**
+   * Synchronize CUDA stream and push a named nvtx range
+   * At the end of the object lifetime, synchronize again and pop the range.
+   *
+   * @param stream stream to synchronize
+   * @param format range name format (accepts printf-style arguments)
+   * @param args the arguments for the printf-style formatting
+   */
+  template <typename... Args>
+  AUTO_RANGE(rmm::cuda_stream_view stream, const char* format, Args... args)
+    : streamMaybe(std::make_optional(stream))
+  {
+    PUSH_RANGE(stream, format, args...);
+  }
+
+  /**
+   * Push a named nvtx range.
+   * At the end of the object lifetime, pop the range back.
+   *
+   * @param format range name format (accepts printf-style arguments)
+   * @param args the arguments for the printf-style formatting
+   */
+  template <typename... Args>
+  AUTO_RANGE(const char* format, Args... args) : streamMaybe(std::nullopt)
+  {
+    PUSH_RANGE(format, args...);
+  }
+
+  ~AUTO_RANGE()
+  {
+    if (streamMaybe.has_value())
+      POP_RANGE(streamMaybe.value());
```

Comment:
Given that you've already checked that `streamMaybe` has a value immediately beforehand, it is safe to use the non-error-checked version of value retrieval:
```suggestion
      POP_RANGE(*streamMaybe);
```

Reply:
Ok! Though I find the `.value()` version somewhat more readable.